File: policy_gradients/tf.py
"""Reference TensorFlow implementation of the policy gradient algorithm."""

# %% imports
import tensorflow as tf
from tensorflow import Tensor
from tensorflow.keras import layers, Sequential, activations, optimizers, Model
from tensorflow_probability import distributions as tfd
import numpy as np
import gym
from typing import Tuple, List


# %% definitions
def construct_policy(env: gym.Env) -> Model:
    """Given an environment, construct a policy."""
    hidden_units = 20
    action_space = env.action_space.n
    return Sequential([
        layers.Dense(hidden_units, activations.relu),
        layers.Dense(action_space, activations.softmax)
        ])


def select_action(policy: Model, state: np.ndarray) -> Tuple[int, Tensor]:
    """
    Select an action using the policy given the curren state.

    Return the action and the log probability of taking that action.
    """
    probs = policy(tf.expand_dims(state, axis=0))
    m = tfd.Categorical(probs=probs)
    action = m.sample()
    return int(action), m.log_prob(action)


def play_episode(policy: Model, env: gym.Env,
                 render=False) -> Tuple[Tensor, Tensor]:
    """
    Play a single episode using the policy in the environment.

    Returns the rewards and log probablities of the action taken
    at each timestep.
    """
    rewards: List[float] = []
    log_probs: List[Tensor] = []
    state = env.reset()
    done = False
    while not done:
        action, log_prob = select_action(policy, state)
        state, reward, done, _ = env.step(action)
        rewards.append(reward)
        log_probs.append(log_prob)
        if render:
            env.render()
    return tf.constant(rewards, dtype=tf.float64), tf.stack(log_probs)


def discount(rewards: Tensor, gamma: float) -> Tensor:
    """Discount the rewards by gamma."""
    discounted = np.zeros_like(rewards)
    running_sum = 0
    for i in reversed(range(len(rewards))):
        running_sum = rewards[i] + gamma * running_sum
        discounted[i] = running_sum
    return tf.constant(discounted)


def normalize(rewards: Tensor) -> Tensor:
    """Normalize rewards to have mean of 0 and std deviation of 1."""
    eps = np.finfo(np.float32).eps.item()
    mean, variance = tf.nn.moments(rewards, axes=0)
    return (rewards - mean) / (tf.sqrt(variance) + eps)


def improve_policy(policy: Model, env: gym.Env,
                   optimizer: optimizers.Optimizer, episodes=100):
    """Improve the policy utilizing the policy gradient algorithm."""
    for _ in range(episodes):
        with tf.GradientTape() as tape:
            rewards, log_probs = play_episode(policy, env)
            returns = normalize(discount(rewards, gamma=0.99))
            policy_loss = sum(-tf.squeeze(log_probs) * returns)
        vars = policy.trainable_variables
        grads = tape.gradient(policy_loss, vars)
        optimizer.apply_gradients(zip(grads, vars))


# %% entry point
env = gym.make("CartPole-v0")

policy = construct_policy(env)

optimizer = optimizers.Adam(learning_rate=1e-2)

sum(play_episode(policy, env, render=True)[0])

improve_policy(policy, env, optimizer, episodes=100)

env.close()

File: policy_gradients/tf.swift
import Python
import TensorFlow

let gym = Python.import("gym")
let env = gym.make("CartPole-v0")

func numpyToTensor(_ numpy: PythonObject) -> Tensor<Float> {
  return Tensor<Float>(Tensor<Double>(numpy: numpy)!).reshaped(to: [1, 4])
}

struct Policy : Layer {
  var l1, l2: Dense<Float>
  
  init(_ env: PythonObject) {
    let stateSpace = Int(env.observation_space.shape[0])!
    let hiddenUnits = 20
    let actionSpace = Int(env.action_space.n)!
    l1 = Dense<Float>(
       inputSize: stateSpace, outputSize: hiddenUnits, activation: relu)
    l2 = Dense<Float>(
       inputSize: hiddenUnits, outputSize: actionSpace, activation: softmax)
  }
  
  @differentiable
  func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
    return l2.applied(to: l1.applied(to: input, in: context), in: context)
  }
}

func choice(_ probs: Tensor<Float>) -> Int32 {
  var roll = Tensor<Float>(randomUniform: [1])[0]
  for i in (0..<probs.shape[1]).reversed() {
    let prob = probs[0][i]
    if roll <= prob {
      return i
    }
    else {
      roll -= prob
    }
  }
  return 0
}

func selectAction(_ policy: Policy, _ state: Tensor<Float>,
                  in context: Context) -> (Int32, Tensor<Float>) {
  let probs = policy.applied(to: state, in: context)
  let action = Int32(choice(probs))
  return (action, log(probs[0][action]))
}

func playEpisode(_ policy: Policy, _ env: PythonObject,
                 in context: Context, render shouldRender: Bool = false)
                 -> (Tensor<Float>, Tensor<Float>) {
  var rewards: [Float] = []
  var logProbs: [Tensor<Float>] = []
  var state = numpyToTensor(env.reset())
  var done = false
  while !done {
    let (action, logProb) = selectAction(policy, state, in: context)
    let (nextState, reward, isDone, _) = env.step(action).tuple4
    rewards.append(Float(reward)!)
    logProbs.append(logProb)
    state = numpyToTensor(nextState)
    done = Bool(isDone)!
  }
  return (Tensor<Float>(rewards), Tensor<Float>(logProbs))
}

func discount(_ rewards: Tensor<Float>, gamma: Float) -> Tensor<Float> {
  var discounted = Tensor<Float>(zeros: rewards.shape)
  var running_sum = Float(0.0)
  for i in (0..<rewards.shape[0]).reversed() {
    running_sum = Float(rewards[i])! + gamma * running_sum
    discounted[i] = Tensor<Float>(running_sum)
  }
  return discounted
}

func std(_ x: Tensor<Float>) -> Tensor<Float> {
  let mean = x.mean()
  return sqrt(pow(x - mean, 2).sum() / Float(x.shape[0]))
}

func normalize(_ rewards: Tensor<Float>) -> Tensor<Float> {
  return (rewards - rewards.mean()) / std(rewards)
}

func improvePolicy(_ policy: inout Policy, _ env: PythonObject,
                   _ optimizer: Adam<Policy, Float>,
                   in context: Context, episodes: Int) {
  for _ in 1...episodes {
    let (rewards, logProbs) = playEpisode(policy, env, in: context)
    let returns = normalize(discount(rewards, gamma: 0.99))
    let gradients = gradient(at: policy) { model -> Tensor<Float> in
      return (-logProbs * returns).sum()
    }
    optimizer.update(&policy.allDifferentiableVariables, along: gradients)
  }
}

var policy = Policy(env)
let optimizer = Adam<Policy, Float>(learningRate: 0.01)
let context = Context(learningPhase: .training)

let (rewards, _) = playEpisode(policy, env, in: context)
rewards.sum()

improvePolicy(&policy, env, optimizer, in: context, episodes: 100)

let (rewards, _) = playEpisode(policy, env, in: context)
rewards.sum()

